Add pybind11 trampoline class for c10d.Store

pietern opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30415 Add pybind11 trampoline class for c10d.Store**

This enables subclassing of c10d.Store and implementing its interface in Python.

Differential Revision: [D18693018](https://our.internmc.facebook.com/intern/diff/D18693018/)

pietern committed:
Add pybind11 trampoline class for c10d.Store

This enables subclassing of c10d.Store and implementing its interface in Python.

Differential Revision: [D18693018](https://our.internmc.facebook.com/intern/diff/D18693018/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -605,6 +658,58 @@ They are used in specifying strategies for reduction collectives, e.g.,
       py::arg("buffer_size"),
       py::call_guard<py::gil_scoped_release>());
 
+  module.def(
```
Why this test is implemented in `init.cpp` instead of in a cpp test file? It looks a little weird to have a test here.

pietern replied:
The test needs a Python class to properly test the trampoline. I don't think we have a separate Python shared object with test-only code, so this is the next best thing. Not ideal, but not too bad either, IMO.

pietern committed:
Update on "Add pybind11 trampoline class for c10d.Store"


This enables subclassing of c10d.Store and implementing its interface in Python.

Differential Revision: [D18693018](https://our.internmc.facebook.com/intern/diff/D18693018/)

[ghstack-poisoned]

mrshenli reviewed (approved):
Took me some time to understand how this works. This LGTM! 

Shall we also add tests for initializing a PG with a PythonStore?

pietern committed:
Update on "Add pybind11 trampoline class for c10d.Store"


This enables subclassing of c10d.Store and implementing its interface in Python.

Differential Revision: [D18693018](https://our.internmc.facebook.com/intern/diff/D18693018/)

[ghstack-poisoned]

pietern commented:
Thanks for the review, @mrshenli. 

Agree it would be good to add a test initializing a process group with this. An issue came up concerning `str` vs `bytes` types which would have been uncovered by doing this (stuffing binary data in a Python string can throw on encoding related issues).

pietern committed:
Update on "Add pybind11 trampoline class for c10d.Store"


This enables subclassing of c10d.Store and implementing its interface in Python.

Differential Revision: [D18693018](https://our.internmc.facebook.com/intern/diff/D18693018/)

[ghstack-poisoned]

mrshenli reviewed (changes requested):
Test failure is real:

```
Nov 26 14:12:11 ======================================================================
Nov 26 14:12:11 FAIL [0.001s]: test_set_get (__main__.PythonStoreTest)
Nov 26 14:12:11 ----------------------------------------------------------------------
Nov 26 14:12:11 Traceback (most recent call last):
Nov 26 14:12:11   File "test_c10d.py", line 362, in test_set_get
Nov 26 14:12:11     c10d._test_python_store(MyPythonStore())
Nov 26 14:12:11   File "test_c10d.py", line 346, in add
Nov 26 14:12:11     self.set(key, bytes(str(new).encode("utf-8")))
Nov 26 14:12:11   File "test_c10d.py", line 333, in set
Nov 26 14:12:11     raise AssertionError("Expected set to be called with string key")
Nov 26 14:12:11 AssertionError: Expected set to be called with string key
```

pietern commented:
A Python 2.7 failure... hmm. I look forward to the day we remove those builds!

pietern committed:
Update on "Add pybind11 trampoline class for c10d.Store"


This enables subclassing of c10d.Store and implementing its interface in Python.

Differential Revision: [D18693018](https://our.internmc.facebook.com/intern/diff/D18693018/)

[ghstack-poisoned]

pietern commented:
Fixed the failure in ed01bcce214aef965b71028d4b55312ee7f9bc68.

pietern commented:
Test failures unrelated. Merging.
